### PR TITLE
Remove max repository component length restriction

### LINKED ISSUE #241: Remove repository component length restriction
As discussed in https://github.com/docker/docker/issues/10392#issuecomment-76869719, let's remove the repository component length restriction and just limit to a max repository length.

### PR DESCRIPTION
**Pull request by ncdc**
Fixes #241

Signed-off-by: Andy Goldstein <agoldste@redhat.com>

### CODE REVIEW

**Comment by ncdc** [comment]
@stevvooe spurious unrelated test failure?

**Comment by dmp42** [comment]
Seems that one is popping from time to time...
Probably need to investigate.

**Comment by ncdc** [comment]
@dmp42 can you kick circle to get it to retest?

**Comment by dmp42** [comment]
Restarted.
For future investigation - log from the failed build: https://gist.github.com/dmp42/7ba754f8985562efc7d3

**Comment by stevvooe** [comment]
@ncdc While we're at it, should we remove `RepositoryNameMaxComponents`? This is also effectively enforced by the maximum number of characters.

**Comment by stevvooe** [comment]
@ncdc Also, let's update the specification (doc/spec/api.md).

**Comment by ncdc** [comment]
> While we're at it, should we remove RepositoryNameMaxComponents? This is also effectively enforced by the maximum number of characters

So there's no longer a need to restrict it to 5?

> Also, let's update the specification (doc/spec/api.md).

Will do, once we settle on `RepositoryNameMaxComponents`. Will do tomorrow - I need a night off :smile:

**Comment by ncdc** [comment]
@stevvooe the spec currently says that a name must have at least 2 components and must be less than 256 characters. The only mention of a 30-character limit is a reference to the repository names as we know of them today. There is also no mention of a maximum of 5 components.

**Comment by stevvooe** [comment]
@jlhawn @shin- PTAL

**Comment by stevvooe** [comment]
LGTM!

**Comment by ncdc** [comment]
@stevvooe did you want to remove the 5 component max?

**Comment by stevvooe** [comment]
@ncdc Please go ahead.

**Comment by ncdc** [comment]
@stevvooe ok, will do

**Comment by ncdc** [comment]
@stevvooe 5 max removed. I can squash if you need.

**Comment by dmcgowan** [comment]
LGTM

File: registry/api/v2/names.go [hunk 1]
```diff
@@ -15,18 +15,10 @@ const (
...
-	// single repository name slash-delimited component
-	RepositoryNameComponentMaxLength = 30
-
 	// RepositoryNameMinComponents is the minimum number of slash-delimited
 	// components that a repository name must have
 	RepositoryNameMinComponents = 1
 
-	// RepositoryNameMaxComponents is the maximum number of slash-delimited
-	// components that a repository name must have
-	RepositoryNameMaxComponents = 5
```
Comment (stevvooe) [inline registry/api/v2/names.go]: `RepositoryNameRegexp` needs to be modified to not restrict this, as well.

Comment (ncdc) [inline registry/api/v2/names.go]: Sorry, forgot to make that change. I knew that was there but with
everything going on today I missed it. I'll get it in later tonight.

**Comment by stevvooe** [comment]
LGTM!

**Pull request merged**